### ISSUE TITLE
Node segfault when trying to query the aerospike client after closing the connection

### ISSUE THREAD

**Issue by plelevier** [aerospike/aerospike-client-nodejs#88]
Hello,

I found that after closing the connection of a aerospike client, trying to querying it may result in segfaults or missing error logs.

Here is a code snippet to reproduce the problem, based on the quickstart example : 

```javascript
var aerospike = require('aerospike');
var status = aerospike.status;


// Connect to the cluster.
var client = aerospike.client({
    hosts: [ { addr: 'localhost', port: 3000 } ]
});

function connect_cb( err, client) {
    if (err.code == status.AEROSPIKE_OK) {
        console.log("Aerospike Connection Success")
    }
}

client.connect(connect_cb);  // connect is synchronous

client.close(); // Closing the client here to reproduce the issue

// Eat the memory
var memoryEater = []
for (var i = 0; i < 9999999; ++i) {
  memoryEater.push({data: 'hello, world'})
}

// The key of the record we are reading.
var key = aerospike.key('test','demo','foo');

var bins = { i: 123, s: "str"}
var metadata = { ttl: 10000, gen: 1}

// write a record to database.
client.put(key, bins, metadata, function(err, key){

    // Read the same record from database
    client.get(key, function(err, rec, meta) {

        // Check for errors
        if ( err.code == status.AEROSPIKE_OK ) {
            // The record was successfully read.
            console.log(rec, meta);
        }
        else {
            // An error occurred
            console.error('error:', err);
        }
    });
});
```

Now, if you try to execute this code multiple times, the following may append : 
- In the worst case scenario, node will segfault : 
```
Aerospike Connection Success
Segmentation fault: 11
```
- In the best case scenario, you'll get the following output (normal behavior) : 
```
Aerospike Connection Success
Aug 31 2015 10:39:41 UTC: ERROR(21611) [put.cc:191] [execute] - Not connected to cluster to put record
Aug 31 2015 10:39:41 UTC: ERROR(21611) [get.cc:108] [prepare] - Key should be an object
Aug 31 2015 10:39:41 UTC: ERROR(21611) [get.cc:161] [execute] - Not connected to Cluster to perform the operation
error: { code: -2,
  message: 'AEROSPIKE_ERR_PARAM',
  func: '',
  file: '../src/main/client/get.cc',
  line: 163 }
```
- Sometimes, you may just be missing error logs : 
```
Aerospike Connection Success
error: { code: -2,
  message: 'AEROSPIKE_ERR_PARAM',
  func: '',
  file: '../src/main/client/get.cc',
  line: 163 }
```

It seems to me that it is related the close method, where the as and log member variables are freed but not initialized back to NULL : 
- https://github.com/aerospike/aerospike-client-nodejs/blob/master/src/main/client/close.cc#L42

Along with the execute methods, where client->as->cluster is checked for NULL and then client->log is used, as seen in this example :
- https://github.com/aerospike/aerospike-client-nodejs/blob/master/src/main/client/get.cc#L160

Also, by reading this : 
- https://github.com/aerospike/aerospike-client-nodejs/blob/master/src/main/client.cc#L85
I feel like freeing memory for `as` and `log` should not be done in the close method.

By replacing the connect / close / memory eating code in the code snippet with : 
```
client.connect(connect_cb);
client.close();
client.connect(connect_cb);
```
I get the same behaviors (segfaults/missing logs, and sometime it's just working), so we can't safely reuse a aerospike client after closing it once, event if we reconnect it

**Comment by GayathriKaliyamoorthy** [comment]
@plelevier 

The client object is not meant to be used after closing the connection. Ideal usage is to use one client object to communicate with Aerospike Cluster. I will enhance the documentation for Close API and state these explicitly.  As for freeing up `as` and `log`, both the structures should be freed in `Close`, otherwise it will be  a memory leak. I will enhance the error checking code to check for NULL and make sure the driver fails gracefully. Thanks for bringing this up. 

Thanks

**Comment by plelevier** [comment]
@GayathriKaliyamoorthy 

Ok, thanks for the feeback. I guess it's related to how v8 is handling memory.
You're right, failing gracefully should be OK but I'm wondering how it can be handled in the best possible way since the logger is no longer available?

**Comment by jhecking** [comment]
I am going to close this issue. As part of the API changes started off in PR #106 the Node.js client now checks the connection status before passing any commands to the C++ client instance. Any commands issued once the connection has been closed will return a client error immediately.

Example:

````js
// issue88.js
const Aerospike = require('aerospike')
Aerospike.connect(function (error, client) {
  if (error) throw error
  var key = new Aerospike.Key('test', 'demo', 'key1')
  client.put(key, {str: 'hello world'}, function (error) {
    if (error) throw error
  })
  client.close()
  client.get(key, function (error, record) {
    if (error) throw error
    console.log(record)
  })
})
````

````
$ node issue88.js

/Users/jhecking/aerospike/aerospike/aerospike-client-nodejs/issue88.js:10
    if (error) throw error
               ^
AerospikeError: Not connected.
    at AerospikeError.Error (native)
    at new AerospikeError (/Users/jhecking/aerospike/aerospike/aerospike-client-nodejs/lib/data_types.js:65:20)
    at Client.DefaultCallbackHandler [as callbackHandler] (/Users/jhecking/aerospike/aerospike/aerospike-client-nodejs/lib/client.js:70:17)
    at nextTickCb (/Users/jhecking/aerospike/aerospike/aerospike-client-nodejs/lib/client.js:1096:10)
    at doNTCallback0 (node.js:419:9)
    at process._tickCallback (node.js:348:13)
    at Client.connect (/Users/jhecking/aerospike/aerospike/aerospike-client-nodejs/lib/client.js:391:18)
    at Aerospike.connect (/Users/jhecking/aerospike/aerospike/aerospike-client-nodejs/lib/aerospike.js:64:10)
    at Object.<anonymous> (/Users/jhecking/aerospike/aerospike/aerospike-client-nodejs/issue88.js:2:11)
    at Module._compile (module.js:435:26)
````

These changes are on master now and will be released with the upcoming client version 2.0.
